python/cudf/cudf/core/udf/udf_kernel_base.py

```diff
@@ -0,0 +1,193 @@
+# Copyright (c) 2025 NVIDIA CORPORATION.
+
+from abc import ABC, abstractmethod
+from contextlib import nullcontext
+
+import numpy as np
+from numba import cuda, typeof
+from numba.core.errors import TypingError
+from numba.np import numpy_support
+from numba.types import CPointer, Poison, Tuple, boolean, int64, void
+
+from cudf.api.types import is_scalar
+from cudf.core.udf.masked_typing import MaskedType
+from cudf.core.udf.nrt_utils import NRTContext, nrt_enabled
+from cudf.core.udf.strings_typing import str_view_arg_handler
+from cudf.core.udf.utils import (
+    _generate_cache_key,
+    _masked_array_type_from_col,
+    _ptx_file,
+    _supported_cols_from_frame,
+    precompiled as kernel_cache,
+)
+from cudf.utils import cudautils
+from cudf.utils._numba import _CUDFNumbaConfig
+from cudf.utils.performance_tracking import _performance_tracking
+
+
+class ApplyKernelBase(ABC):
+    """
+    Base class for kernels computing the result of `.apply`
+    operations on Series or DataFrame objects.
+    """
+
+    def __init__(self, frame, func, args):
+        self.frame = frame
+        self.func = func
+
+        if not all(is_scalar(arg) for arg in args):
+            raise TypeError("only scalar valued args are supported by apply")
+        self.args = args
+        self.nrt = False
+
+        self.frame_type = self._get_frame_type()
+        self.device_func = cuda.jit(device=True)(self.func)
+
+    @property
+    @abstractmethod
+    def kernel_type(self):
+        """
+        API type launching the kernel, used to break
+        degenerecies in the cache.
+        """
+
+    @abstractmethod
+    def _get_frame_type(self):
+        """
+        Numba type of the frame being passed to the kernel.
+        """
+
+    @abstractmethod
+    def _get_kernel_string(self):
+        """
+        Generate executable string of python that defines a
+        kernel we may retrieve from the context and compile
+        with numba.
+        """
+
+    @abstractmethod
+    def _get_kernel_string_exec_context(self):
```

Comment:
Can we mark this method as cached (in the subclasses)? I see that we do currently modify the resulting dict directly with `compile_kernel_string` where we're using this as a context for calling `exec`, but from a quick look it seems like the context should be safe to reuse since we overwrite the function. Please ignore this if I'm wrong, though.

Reply:
Done